operator/builtin/input/file/file.go

```diff
@@ -111,55 +118,67 @@ func (f *InputOperator) startPoller(ctx context.Context) {
-			// This is done here because the next generation is about to start
-			for i := 0; i < len(f.knownFiles); i++ {
-				f.knownFiles[i].generation++
-			}
-
-			// Get the list of paths on disk
-			matches = getMatches(f.Include, f.Exclude)
-			if f.firstCheck && len(matches) == 0 {
-				f.Warnw("no files match the configured include patterns", "include", f.Include)
-			} else if len(matches) > f.MaxConcurrentFiles {
-				matches, f.queuedMatches = matches[:f.MaxConcurrentFiles], matches[f.MaxConcurrentFiles:]
-			}
-		}
+	matches := getMatches(f.Include, f.Exclude)
+	if f.firstCheck && len(matches) == 0 {
+		f.Warnw("no files match the configured include patterns", "include", f.Include)
 	}
-
-	// Open the files first to minimize the time between listing and opening
-	files := make([]*os.File, 0, len(matches))
 	for _, path := range matches {
-		if _, ok := f.SeenPaths[path]; !ok {
-			if f.startAtBeginning {
-				f.Infow("Started watching file", "path", path)
-			} else {
-				f.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
-			}
-			f.SeenPaths[path] = struct{}{}
-		}
-		file, err := os.Open(path)
+		reader, err := f.isNewFileToTail(path)
 		if err != nil {
-			f.Errorw("Failed to open file", zap.Error(err))
 			continue
 		}
-		files = append(files, file)
-	}
+		if reader == nil {
+			continue
+		}
 
-	readers := f.makeReaders(files)
+		if _, ok := f.tailingReaders[path]; !ok { // If path doesn't exist => new file to tail
+			if f.firstCheck {
+				if f.startAtBeginning {
+					f.Infow("Started watching file", "path", path)
+				} else {
+					f.Infow("Started watching file from end. To read preexisting logs, configure the argument 'start_at' to 'beginning'", "path", path)
+				}
+				if err := reader.InitializeOffset(f.startAtBeginning); err != nil {
+					f.Errorw("Failed to initialize offset for "+path, zap.Error(err))
+				}
+			} else {
+				f.Infow("Started watching file", "path", path)
+			}
+			f.tailingReaders[path] = []*Reader{reader}
+		} else { // If path exists
+			f.Infow("Log rotation detected. Started watching file", "path", path)
+			f.tailingReaders[path] = append(f.tailingReaders[path], reader)
+			// Limit tracking the iteration of rotated files
+			if len(f.tailingReaders[path]) > RotatedFileTrackingLimit {
+				var firstReader *Reader
+				firstReader, f.tailingReaders[path] = f.tailingReaders[path][0], f.tailingReaders[path][1:]
+				f.closeFile(firstReader.file, firstReader.Path)
+			}
+		}
+	}
 	f.firstCheck = false
 
+	readerCount := 0
+	for _, readers := range f.tailingReaders {
+		readerCount += len(readers)
+	}
+
+	if len(f.readerQueue) <= readerCount {
+		for _, readers := range f.tailingReaders {
+			f.readerQueue = append(f.readerQueue, readers...)
+		}
+	}
+
+	count := min0(f.MaxConcurrentFiles, readerCount)
+	count = min0(count, len(f.readerQueue))
+	polledReaders := f.readerQueue[:count]
```

Comment:
Unfortunately, I don't think `max_concurrent_files` is really being respected with this design. 

The original problem that this setting was introduced for, is when there are so many matching files that we run out of file descriptors. By leaving all the files open, we are not limiting the number of file descriptors.

I'm not certain that it's possible to accomplish both objectives though (limiting file descriptors and tracking rotations), so perhaps we would need to set a reasonably high default for `max_concurrent_files` and then only keep that many open. (This would mean that if you are matching too many files, you may lose a small percent of logs upon rotation)

Reply:
Obviously not an ideal situation, but I think users must be able to allocate a fixed number of file descriptors to the receiver.

Reply:
you are right. so set it to reasonably high like like 5,000? we definitely need to have a workflow to close out already rotated files to be support this. but even then, in situation where there are 6000 actively written files and maximum is 5000, then it will only tail 5000 files and not tail the rest 1000 files at all.